language/move-binary-format/src/control_flow_graph.rs

```diff
@@ -112,56 +107,63 @@ impl VMControlFlowGraph {
         let blocks = blocks;
         assert_eq!(entry, code_len);
 
-        // Determine traversal order
-        // build a DAG subgraph (remove the loop back edges)
-        let dag: Map<BlockId, Set<BlockId>> = blocks
-            .iter()
-            .map(|(id, block)| {
-                let id = *id;
-                let non_loop_continue_successors = block
-                    .successors
-                    .iter()
-                    // remove the loop back edges
-                    // this simple check for back edges relies on guarantees from the control flow
-                    // verifier (control_flow.rs)
-                    .filter(|successor| !is_back_edge(id, **successor))
-                    .copied()
-                    .collect();
-                (id, non_loop_continue_successors)
-            })
-            .collect();
-        // assert it is a dag
-        debug_assert!(dag.iter().all(|(id, successors)| successors
-            .iter()
-            .all(|successor| !is_back_edge(*id, *successor))));
+        #[derive(Copy, Clone)]
+        enum Exploration {
+            InProgress,
+            Done,
+        }
 
-        // build the post-order traversal
+        let mut loop_heads: Map<BlockId, Set<BlockId>> = Map::new();
+        let mut exploration: Map<BlockId, Exploration> = Map::new();
+        let mut stack = vec![ENTRY_BLOCK_ID];
         let mut post_order = Vec::with_capacity(blocks.len());
-        let mut finished = Set::new();
-        let mut stack = vec![(ENTRY_BLOCK_ID, /* is_first_visit */ true)];
-        while let Some((cur, is_first_visit)) = stack.pop() {
-            if is_first_visit {
-                stack.push((cur, false));
-                stack.extend(
-                    dag[&cur]
-                        .iter()
-                        .filter(|successor| !finished.contains(*successor))
-                        .map(|successor| (*successor, /* is_first_visit */ true)),
-                );
-            } else {
-                debug_assert!(dag[&cur]
-                    .iter()
-                    .all(|successor| finished.contains(successor)));
-                if finished.insert(cur) {
-                    post_order.push(cur)
+
+        while let Some(block) = stack.pop() {
+            match exploration.entry(block) {
+                Entry::Vacant(entry) => {
+                    entry.insert(Exploration::InProgress);
+
+                    // Push the block back on the stack to finish processing it once the sub-graph
+                    // reachable from it has been traversed.
+                    stack.push(block);
+
+                    for succ in &blocks[&block].successors {
+                        match exploration.get(succ) {
+                            // Frontier detected
+                            None => stack.push(*succ),
+
+                            // Back edge detected
+                            Some(Exploration::InProgress) => {
```

Comment:
This intuitively makes sense to me. But is there some way we can sort of formalize the overall approach here in a comment block maybe. 

My proposed description, which might help check my understanding: "In this DFS approach, if you encounter an `Exploration::Done` node, then it means the node was visited via a different path, meaning there is a 'diamond' in the CFG. If however, you encounter an `Exploration:InProgress` node, then it means that one of the blocks predecessors is also a successor, meaning there is a 'loop' in the CFG. We greedily take this edge as the 'back-edge' in the loop"

Reply:
Though it might be nice to note why you cannot possibly hit a `Done` in a loop, and why you cannot hit an `InProgress` in a diamond. I think it is at least quickly dismissed via contradiction, but if you have something constructive in mind, might be helpful too

Reply:
I'm interpreting your question mainly as being related to loop heads, where the formal properties are that:

- For every loop in the CFG (that is reachable from the entry block), we assign at least one back edge (i.e. back edge if and only if loop)
- In reducible CFGs, the node that the back edge points to dominates all the other nodes in the loop (that's what makes it the head).

The first property's forward direction (back edge => loop) is explained by your description, for loop => back edge,  the sketch is: 

- Suppose there is a loop and it's reachable from the entry block.
- Consider the point at which we start to explore the first node in that loop, `F` as part of the traversal (that node is now in-progress, everything else is not present in the exploration).
- Because `F` is part of a loop, it will have an edge pointing to it, from a node `L` where there is a path from `F` to `L` as well.
- `F` will not transition to `Done` until all the nodes reachable from it (including `L`) have been visited, 
- Because `F` is the first node we are seeing in the loop, this means all the other nodes in the loop, (again, including `L`) will be visited while `F` is `in-progress`.
- Therefore we will process the `L -> F` edge while `F` is in progress.
- Therefore all loops will get a back edge.

The property that all reducible CFGs correctly identify a sensible loop head (the node that dominates the others in the loop) is a bit more fiddly, and it's covered in the Bourdoncle paper, so I may just refer to that... (this property comes from the fact that we are using a DFS from the entry block and reducible CFGs are defined by this "domination" property)

Reply:
We should shove some condensed version of my comment and your description into this file. And point to the Bourdoncle paper. Luckily though, this section is far less critical than #579, so we should focus an heavy documents to that PR/verifier pass.